### PR TITLE
Optimize s.c.m.Set#filterInPlace implementations

### PR DESCRIPTION
**Pull request by joshlemer**
inlcuding:
* s.c.m.HashSet
* s.c.m.LinkedHashSet
* JSetWrapper forwards to underlying.removeIf
* ~forward MutableJsetWrapper#removeIf to underlying.filterInPlace~
* Add preemptive overrides for mutable.TreeSet#filterInPlace and filterImpl
* s.c.m.LinkedHashMap

## TODO

- [x] Create some good benchmarks to confirm that this actually is beneficial

## Benchmarks

* Turned out that it's slower to forward JSetWrapper#removeIf to underlying.filterInPlace
* Everything else is ~40% faster or so, ~except for very large (1million+) sized LinkedHashMaps. I'm not sure why at large sizes the old beats the new one, something to look into. Still, probably worth it for the great speedups at smaller sizes I'd say~. Fixed it, now it's faster at all sizes (benchmarks appended below)

```
Set Benchmarks

HashSet

[info] Benchmark                              (size)  Mode  Cnt         Score         Error  Units
[info] SetBenchmark.hashSetFilterInPlaceNew        0  avgt    6        20.642 ±       0.570  ns/op
[info] SetBenchmark.hashSetFilterInPlaceNew        1  avgt    6        37.340 ±       0.517  ns/op
[info] SetBenchmark.hashSetFilterInPlaceNew       10  avgt    6        74.699 ±       1.608  ns/op
[info] SetBenchmark.hashSetFilterInPlaceNew      100  avgt    6       719.398 ±      28.625  ns/op
[info] SetBenchmark.hashSetFilterInPlaceNew     1000  avgt    6      6081.337 ±     916.343  ns/op
[info] SetBenchmark.hashSetFilterInPlaceNew  1000000  avgt    6  10756491.709 ± 6216937.005  ns/op
[info] SetBenchmark.hashSetFilterInPlaceOld        0  avgt    6        22.983 ±       0.787  ns/op
[info] SetBenchmark.hashSetFilterInPlaceOld        1  avgt    6        58.709 ±       1.616  ns/op
[info] SetBenchmark.hashSetFilterInPlaceOld       10  avgt    6       200.766 ±       5.058  ns/op
[info] SetBenchmark.hashSetFilterInPlaceOld      100  avgt    6      1314.876 ±      19.072  ns/op
[info] SetBenchmark.hashSetFilterInPlaceOld     1000  avgt    6     10476.401 ±     167.688  ns/op
[info] SetBenchmark.hashSetFilterInPlaceOld  1000000  avgt    6  16159382.390 ± 3181990.396  ns/op

LinkedHashSet

[info] Benchmark                          (size)  Mode  Cnt         Score          Error  Units
[info] SetBenchmark.lhsFilterInPlaceNew        0  avgt    6        21.562 ±        0.163  ns/op
[info] SetBenchmark.lhsFilterInPlaceNew        1  avgt    6        40.113 ±        2.406  ns/op
[info] SetBenchmark.lhsFilterInPlaceNew       10  avgt    6        80.435 ±        2.422  ns/op
[info] SetBenchmark.lhsFilterInPlaceNew      100  avgt    6       618.553 ±       10.477  ns/op
[info] SetBenchmark.lhsFilterInPlaceNew     1000  avgt    6     14805.856 ±      374.524  ns/op
[info] SetBenchmark.lhsFilterInPlaceNew  1000000  avgt    6  79483415.292 ±  2776669.384  ns/op
[info] SetBenchmark.lhsFilterInPlaceOld        0  avgt    6        21.479 ±        0.104  ns/op
[info] SetBenchmark.lhsFilterInPlaceOld        1  avgt    6       119.541 ±        9.858  ns/op
[info] SetBenchmark.lhsFilterInPlaceOld       10  avgt    6       235.250 ±        7.125  ns/op
[info] SetBenchmark.lhsFilterInPlaceOld      100  avgt    6      2136.341 ±       66.079  ns/op
[info] SetBenchmark.lhsFilterInPlaceOld     1000  avgt    6     17373.401 ±      193.661  ns/op
[info] SetBenchmark.lhsFilterInPlaceOld  1000000  avgt    6  61029471.158 ± 25085732.002  ns/op

JSetWrapper

[info] Benchmark                                  (size)  Mode  Cnt         Score         Error  Units
[info] SetBenchmark.jsetWrapperFilterInPlaceNew        0  avgt    6        23.334 ±       0.279  ns/op
[info] SetBenchmark.jsetWrapperFilterInPlaceNew        1  avgt    6        43.792 ±       0.612  ns/op
[info] SetBenchmark.jsetWrapperFilterInPlaceNew       10  avgt    6       107.306 ±       4.030  ns/op
[info] SetBenchmark.jsetWrapperFilterInPlaceNew      100  avgt    6       932.517 ±      11.078  ns/op
[info] SetBenchmark.jsetWrapperFilterInPlaceNew     1000  avgt    6      7812.614 ±     159.583  ns/op
[info] SetBenchmark.jsetWrapperFilterInPlaceNew  1000000  avgt    6  10011237.971 ±  268929.575  ns/op
[info] SetBenchmark.jsetWrapperFilterInPlaceOld        0  avgt    6        24.300 ±       0.213  ns/op
[info] SetBenchmark.jsetWrapperFilterInPlaceOld        1  avgt    6        79.599 ±       2.664  ns/op
[info] SetBenchmark.jsetWrapperFilterInPlaceOld       10  avgt    6       285.031 ±      16.937  ns/op
[info] SetBenchmark.jsetWrapperFilterInPlaceOld      100  avgt    6      1938.743 ±      18.156  ns/op
[info] SetBenchmark.jsetWrapperFilterInPlaceOld     1000  avgt    6     20505.604 ±     520.052  ns/op
[info] SetBenchmark.jsetWrapperFilterInPlaceOld  1000000  avgt    6  21605549.296 ± 5291256.367  ns/op


SetWrapper

// !!! The below benchmarks determined the old implementation was faster, so that change has been reverted!

[info] Benchmark                                   (size)  Mode  Cnt         Score         Error  Units
[info] SetBenchmark.mutableSetWrapperRemoveIfNew        0  avgt    6        21.505 ±       0.308  ns/op
[info] SetBenchmark.mutableSetWrapperRemoveIfNew        1  avgt    6        51.029 ±       0.893  ns/op
[info] SetBenchmark.mutableSetWrapperRemoveIfNew       10  avgt    6       134.889 ±       1.601  ns/op
[info] SetBenchmark.mutableSetWrapperRemoveIfNew      100  avgt    6      1353.354 ±      27.763  ns/op
[info] SetBenchmark.mutableSetWrapperRemoveIfNew     1000  avgt    6     15635.084 ±    9842.301  ns/op
[info] SetBenchmark.mutableSetWrapperRemoveIfNew  1000000  avgt    6  11980124.763 ± 2516868.853  ns/op
[info] SetBenchmark.mutableSetWrapperRemoveIfOld        0  avgt    6        33.541 ±       2.788  ns/op
[info] SetBenchmark.mutableSetWrapperRemoveIfOld        1  avgt    6        41.562 ±       1.162  ns/op
[info] SetBenchmark.mutableSetWrapperRemoveIfOld       10  avgt    6       108.295 ±       3.015  ns/op
[info] SetBenchmark.mutableSetWrapperRemoveIfOld      100  avgt    6       881.835 ±      44.200  ns/op
[info] SetBenchmark.mutableSetWrapperRemoveIfOld     1000  avgt    6      7853.866 ±     175.839  ns/op
[info] SetBenchmark.mutableSetWrapperRemoveIfOld  1000000  avgt    6   9396194.135 ±  575615.688  ns/op

```

```
LinkedHashSet

[info] Benchmark                          (size)  Mode  Cnt          Score          Error  Units
[info] SetBenchmark.lhsFilterInPlaceNew        0  avgt    6         21.896 ±        0.513  ns/op
[info] SetBenchmark.lhsFilterInPlaceNew        1  avgt    6         35.793 ±        0.313  ns/op
[info] SetBenchmark.lhsFilterInPlaceNew       10  avgt    6         74.467 ±        2.220  ns/op
[info] SetBenchmark.lhsFilterInPlaceNew      100  avgt    6        545.094 ±       19.071  ns/op
[info] SetBenchmark.lhsFilterInPlaceNew     1000  avgt    6       9972.029 ±      308.070  ns/op
[info] SetBenchmark.lhsFilterInPlaceNew  1000000  avgt    6   58538950.667 ±  3206886.603  ns/op
[info] SetBenchmark.lhsFilterInPlaceNew  2000000  avgt    6  120526442.167 ±  8950859.500  ns/op
[info] SetBenchmark.lhsFilterInPlaceOld        0  avgt    6         22.343 ±        0.899  ns/op
[info] SetBenchmark.lhsFilterInPlaceOld        1  avgt    6        113.013 ±        4.684  ns/op
[info] SetBenchmark.lhsFilterInPlaceOld       10  avgt    6        227.760 ±        6.784  ns/op
[info] SetBenchmark.lhsFilterInPlaceOld      100  avgt    6       1878.171 ±       70.605  ns/op
[info] SetBenchmark.lhsFilterInPlaceOld     1000  avgt    6      17791.455 ±     1588.061  ns/op
[info] SetBenchmark.lhsFilterInPlaceOld  1000000  avgt    6   63002982.083 ± 21877377.837  ns/op
[info] SetBenchmark.lhsFilterInPlaceOld  2000000  avgt    6  151673581.250 ± 85351947.209  ns/op
```

### CODE REVIEW

File: src/library/scala/collection/mutable/LinkedHashMap.scala [hunk 1]
```diff
@@ -129,6 +132,59 @@ class LinkedHashMap[K, V]
...
+      if (earlier ne null) {
+        // Rather than set each node's `later` pointer to `next.next` as we iterate, we wait until we find the next
+        // node that passes the filter, then go back to `earlier` and set it's `later` to that node.
+        // However that means the last node that passes the filter will still point to its previous tail (if it wasn't
+        // the last element of the Map). Now at the end we make sure to cut that link.
+        earlier.later = null
+      }
+    }
+    this
+  }
```
Comment (joshlemer) [inline src/library/scala/collection/mutable/LinkedHashMap.scala:180]: Actually I think there's some logical issues here.. consider this WIP

File: src/library/scala/collection/convert/Wrappers.scala [hunk 2]
```diff
@@ -214,6 +215,15 @@ private[collection] trait Wrappers {
       new JSetWrapper[A](new ju.LinkedHashSet[A](underlying))
 
     override def iterableFactory = mutable.HashSet
+
+    override def filterInPlace(p: A => Boolean): this.type = {
+      if (underlying.size() > 0) {
+        underlying.removeIf(new Predicate[A] {
```
Comment (Ichoran) [inline src/library/scala/collection/convert/Wrappers.scala]: This is a SAM, right?  Can't you just `a => !p(a)` instead of the whole `new Predicate` mess?

File: src/library/scala/collection/convert/Wrappers.scala [hunk 3]
```diff
@@ -187,8 +187,9 @@ private[collection] trait Wrappers {
     }
     override def remove(elem: AnyRef) =
       try underlying.remove(elem.asInstanceOf[A])
-      catch { case ex: ClassCastException => false }
-    override def clear() = underlying.clear()
+      catch { case _: ClassCastException => false }
+
+    override def clear(): Unit = underlying.clear()
```
Comment (Ichoran) [inline src/library/scala/collection/convert/Wrappers.scala]: Have I mentioned lately how much I hate `foo() = unitReturningThing`?  Thanks for catching this one!

File: src/library/scala/collection/mutable/HashTable.scala [hunk 4]
```diff
@@ -131,6 +131,12 @@ private[mutable] abstract class HashTable[A, B, Entry >: Null <: HashEntry[A, En
     foreachEntry(writeEntry)
   }
 
+  /** Get the entry at a given index, or null if none exist */
+  final def apply(index: Int): Entry = table(index).asInstanceOf[Entry]
+
+  /** Set the entry at bucket `index` to `entry` */
+  final def update(index: Int, entry: Entry): Unit = table(index) = entry
```
Comment (Ichoran) [inline src/library/scala/collection/mutable/HashTable.scala]: Argh, another reason to hate `: Unit =`.  Can you add some braces so we don't get `foo: X = x = y` (and have `foo: X = { x = y }` instead?  If you are a compiler it makes perfect sense but not everyone is a compiler.

File: src/library/scala/collection/mutable/HashSet.scala [hunk 5]
```diff
@@ -203,6 +202,41 @@ final class HashSet[A](initialCapacity: Int, loadFactor: Double)
...
   }
 
+  override def filterInPlace(p: A => Boolean): HashSet.this.type = {
+    if (isEmpty) {
+      this
+    } else {
+      var bucket = 0
+
+      while (bucket < table.length) {
+        table(bucket) = table(bucket) match {
```
Comment (Ichoran) [inline src/library/scala/collection/mutable/HashSet.scala]: Matching out `null` creates extra indentation for not much benefit.  Can you rewrite it as `val head = table(bucket)` with a conditional if it's not null?

File: src/library/scala/collection/mutable/HashSet.scala [hunk 6]
```diff
@@ -203,6 +202,41 @@ final class HashSet[A](initialCapacity: Int, loadFactor: Double)
     }
   }
 
+  override def filterInPlace(p: A => Boolean): HashSet.this.type = {
+    if (isEmpty) {
+      this
```
Comment (Ichoran) [inline src/library/scala/collection/mutable/HashSet.scala]: Just do one `if (nonEmpty)` and leave `this` at the very end (outside of the condition).  Cleaner.

File: src/library/scala/collection/mutable/HashSet.scala [hunk 7]
```diff
@@ -203,6 +202,41 @@ final class HashSet[A](initialCapacity: Int, loadFactor: Double)
...
+
+      while (bucket < table.length) {
+        table(bucket) = table(bucket) match {
+          case null => null
+          case head =>
+            var newHead: Node[A] = null
+            var prev: Node[A] = null
+            var next = head
+            while (next ne null) {
+              if (p(next.key)) {
```
Comment (Ichoran) [inline src/library/scala/collection/mutable/HashSet.scala]: This logic is unnecessarily hard to follow with all the braces.  It's actually very simple, but can we keep the one-liners on one line instead of three?

File: src/library/scala/collection/mutable/HashSet.scala [hunk 8]
```diff
@@ -203,6 +202,41 @@ final class HashSet[A](initialCapacity: Int, loadFactor: Double)
...
+                if (newHead eq null) {
```
Comment (Ichoran) [inline src/library/scala/collection/mutable/HashSet.scala]: I think it would be faster yet if you split the loop in two; make `head` a var, and first do `while ((head ne null) && !p(head.key)) head = head.next`.  Then you can do the second part of the loop without having to check for `newHead` and stuff.

File: src/library/scala/collection/mutable/LinkedHashMap.scala [hunk 9]
```diff
@@ -129,6 +132,59 @@ class LinkedHashMap[K, V]
...
+    if (size > 0) {
+      var bucket = 0 // the index of the current bucket in the hash table
+      val tableTable = table.table
+      val buckets = tableTable.length
+      var newFirstEntry: Entry = null // will become the new firstEntry
+      var earlier: Entry = null // the entry "before" the current entry, may cross bucket boundaries
+      var newTableSize = table.tableSize
+
+      while (bucket < buckets) {
+        (tableTable(bucket) : @switch) match {
```
Comment (Ichoran) [inline src/library/scala/collection/mutable/LinkedHashMap.scala]: I don't understand why you're switching on a match statement instead of using a simple conditional for the non-null case.

File: src/library/scala/collection/mutable/LinkedHashMap.scala [hunk 10]
```diff
@@ -129,6 +132,59 @@ class LinkedHashMap[K, V]
...
+          case null => ()
+          case head =>
+            var newHead: Entry = null // the new beginning of this bucket
+          var prev: Entry = null // the previous node within this bucket, does not cross bucket boundaries
+          var next: Entry = head.asInstanceOf[Entry] // the next node within this bucket, does not cross bucket boundaries
+
+            while (next ne null) {
```
Comment (Ichoran) [inline src/library/scala/collection/mutable/LinkedHashMap.scala]: I think this also should be split into prefix and postfix loops as I described for `HashMap`.

File: src/library/scala/collection/mutable/LinkedHashMap.scala [hunk 11]
```diff
@@ -129,6 +132,59 @@ class LinkedHashMap[K, V]
...
+      }
+      table.tableSize = newTableSize
+      firstEntry = newFirstEntry
+      lastEntry = earlier
+      if (earlier ne null) {
+        // Rather than set each node's `later` pointer to `next.next` as we iterate, we wait until we find the next
+        // node that passes the filter, then go back to `earlier` and set it's `later` to that node.
+        // However that means the last node that passes the filter will still point to its previous tail (if it wasn't
+        // the last element of the Map). Now at the end we make sure to cut that link.
+        earlier.later = null
```
Comment (Ichoran) [inline src/library/scala/collection/mutable/LinkedHashMap.scala]: This doesn't sound right.  You're going through things in hash order, not insertion order, right?  So you have got to heal the pointers as you go.

File: src/library/scala/collection/mutable/LinkedHashSet.scala [hunk 12]
```diff
@@ -95,6 +95,64 @@ class LinkedHashSet[A]
     }
   }
 
+  override def filterInPlace(p: A => Boolean): LinkedHashSet.this.type = {
+    if (size > 0) {
+      if (size < 1000000) {
```
Comment (Ichoran) [inline src/library/scala/collection/mutable/LinkedHashSet.scala]: When you improve the `LinkedHashMap` code, I bet the 1M thing will go away.  So, probably, you'll just need to redo this all to match LHM.

File: src/library/scala/collection/mutable/Set.scala [hunk 13]
```diff
@@ -78,8 +78,14 @@ trait SetOps[A, +CC[X], +C <: SetOps[A, CC, C]]
     *             are removed.
     */
   def filterInPlace(p: A => Boolean): this.type = {
-    for (elem <- this.toList) // scala/bug#7269 toList avoids ConcurrentModificationException
-      if (!p(elem)) this -= elem
+    var list = this.toList // scala/bug#7269 toList avoids ConcurrentModificationException
```
Comment (Ichoran) [inline src/library/scala/collection/mutable/Set.scala]: If we know the size--and we should--using `toArray` will avoid a huge batch of garbage.

**Review by Ichoran — CHANGES_REQUESTED** [review]
Lots of changes here--do you have time?  The goal is admirable, and the code is decent, but I don't think it's entirely trivial to get it into merge-worthy shape.

File: src/library/scala/collection/mutable/LinkedHashMap.scala (hunk 11 above)
Comment (joshlemer) [reply src/library/scala/collection/mutable/LinkedHashMap.scala]: You're absolutely right.

File: src/library/scala/collection/mutable/HashSet.scala (hunk 5 above)
Comment (joshlemer) [reply src/library/scala/collection/mutable/HashSet.scala]: :+1:

File: src/library/scala/collection/mutable/LinkedHashMap.scala (hunk 10 above)
Comment (joshlemer) [reply src/library/scala/collection/mutable/LinkedHashMap.scala]: :+1:

File: src/library/scala/collection/mutable/Set.scala (hunk 13 above)
Comment (joshlemer) [reply src/library/scala/collection/mutable/Set.scala]: Creating an Array requires a classtag though, I've now created an Array[Any] but that ill box for primitives

File: src/library/scala/collection/mutable/HashSet.scala (hunk 6 above)
Comment (joshlemer) [reply src/library/scala/collection/mutable/HashSet.scala]: :+1:

**Comment by joshlemer** [comment]
I can confirm that it's faster to use arrays than lists like @Ichoran said, for the default implementation of mutable.Set.filterInPlace, so I've made that change:

```
[info] Benchmark                              (setType)   (size)  Mode  Cnt         Score         Error  Units
[info] SetBenchmark2.filterInPlace_old_List     HashSet        0  avgt   10        21.765 ±       0.277  ns/op
[info] SetBenchmark2.filterInPlace_old_List     HashSet        1  avgt   10       103.894 ±       1.399  ns/op
[info] SetBenchmark2.filterInPlace_old_List     HashSet       10  avgt   10       378.026 ±     242.048  ns/op
[info] SetBenchmark2.filterInPlace_old_List     HashSet      100  avgt   10      1826.028 ±     167.992  ns/op
[info] SetBenchmark2.filterInPlace_old_List     HashSet     1000  avgt   10     14026.203 ±     449.833  ns/op
[info] SetBenchmark2.filterInPlace_old_List     HashSet    10000  avgt   10    146571.122 ±   30432.623  ns/op
[info] SetBenchmark2.filterInPlace_old_List     HashSet   100000  avgt   10   1431094.013 ±   59834.840  ns/op
[info] SetBenchmark2.filterInPlace_old_List     HashSet  1000000  avgt   10  17710727.556 ± 1853070.105  ns/op
[info] SetBenchmark2.filterInPlace_old_array    HashSet        0  avgt   10        22.323 ±       1.018  ns/op
[info] SetBenchmark2.filterInPlace_old_array    HashSet        1  avgt   10        51.926 ±       1.487  ns/op
[info] SetBenchmark2.filterInPlace_old_array    HashSet       10  avgt   10       175.429 ±      11.073  ns/op
[info] SetBenchmark2.filterInPlace_old_array    HashSet      100  avgt   10      1110.138 ±      96.133  ns/op
[info] SetBenchmark2.filterInPlace_old_array    HashSet     1000  avgt   10      9631.570 ±     199.303  ns/op
[info] SetBenchmark2.filterInPlace_old_array    HashSet    10000  avgt   10     90759.773 ±    4281.902  ns/op
[info] SetBenchmark2.filterInPlace_old_array    HashSet   100000  avgt   10    896326.958 ±   81138.864  ns/op
[info] SetBenchmark2.filterInPlace_old_array    HashSet  1000000  avgt   10  10879596.631 ±  383761.516  ns/op

```

**Comment by joshlemer** [comment]
I think the best way forward for this work is to split it out into different PR's for each implementation so that review is more manageable and some good implementations aren't held back by others.
